File: Java/1416. Restore The Array.java
class Solution {
    private int[] dp;
    private int m, mod = (int)Math.pow(10, 9) + 7;

    public int numberOfArrays(String s, int k) {
        m = s.length();
        dp = new int[m + 1];
        return dfs(0, s, k) % mod;
    }

    private int dfs(int start, String s, int k) {
        if (dp[start] > 0) {
            return dp[start];
        }

        if (start == m) {
            return 1;
        }

        if (s.charAt(start) == '0') {
            return 0;
        }

        int count = 0;
        for (int end=start; end<m; end++) {
            Long cur = Long.parseLong(s.substring(start, end+1));
            if (cur > k) {
                break;
            }
            count =(count + dfs(end + 1, s, k)) % mod;
        }
        dp[start] = count;
        return count;
    }
}

File: Python3/1416. Restore The Array.py
class Solution:
    def numberOfArrays(self, s: str, k: int) -> int:
        m, n = len(s), len(str(k))
        mod = 10**9 + 7
        dp = [0] * (m + 1)

        def dfs(start):
            # we have already computed dp[start]
            if dp[start] != 0:
                return dp[start]

            # reach to the end of s
            if start == m:
                return 1

            # current is '0', which cannot be a breakpoint
            if s[start] == '0':
                return 0

            count = 0
            for end in range(start, m):
                cur = s[start: end + 1]
                if int(cur) > k:
                    break
                count += dfs(end + 1)
            
            dp[start] = count % mod
            return count

        return dfs(0) % mod